### PR TITLE
[FIX] base, web: error handling of invalid leaf

### PR DESCRIPTION
**Pull request by sauc-odoo**
Currently, when an invalid domain causes an error it raises an Exception which will be caught by the sentry and generates unnecessary traffic.

So, we stop catching exceptions from invalid domains that are supposed to be generated by User's Mistakes.

Steps to reproduce:
    1. Turn on developer mode.
    2. Install mass_mailing module and open it.
    3. Click on the new button.
    4. Enter subject and set recipients as contact or mailing contact.
    5. In code editor pass the domain as [('', '=', 0)]
    6. error will occur - ValueError: Invalid leaf ('', '=', 0)

See this traceback: 
```
ValueError: Invalid leaf ('', '=', 1)
  File "odoo/http.py", line 2115, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1698, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 134, in retrying
    result = func()
  File "odoo/http.py", line 1725, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1922, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "addons/website/models/ir_http.py", line 234, in _dispatch
    response = super()._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 715, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web/controllers/dataset.py", line 28, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "addons/web/controllers/dataset.py", line 24, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "odoo/api.py", line 457, in call_kw
    result = _call_kw_model(method, model, args, kwargs)
  File "odoo/api.py", line 430, in _call_kw_model
    result = method(recs, *args, **kwargs)
  File "odoo/models.py", line 1485, in search_count
    query = self._search(domain, limit=limit)
  File "odoo/addons/base/models/res_partner.py", line 899, in _search
    return super()._search(domain, offset, limit, order, access_rights_uid)
  File "odoo/models.py", line 4759, in _search
    query = self._where_calc(domain)
  File "odoo/models.py", line 4506, in _where_calc
    return expression.expression(domain, self).query
  File "odoo/osv/expression.py", line 448, in __init__
    self.parse()
  File "odoo/osv/expression.py", line 613, in parse
    push(leaf, self.root_model, self.root_alias)
  File "odoo/osv/expression.py", line 600, in push
    check_leaf(leaf, internal)
  File "odoo/osv/expression.py", line 397, in check_leaf
    raise ValueError("Invalid leaf %s" % str(element))
```

Applying this commit will fix this issue.

sentry-3945982195

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -1468,6 +1468,14 @@ def user_has_groups(self, groups):
 
         return not has_groups
 
+    @api.model
+    def check_domain(self, domain):
+        try:
+            return self.search_count(domain)
+        except Exception as e:
+            e.sentry_ignored = True
+            raise
```
Comment (yelizariev) [inline odoo/models.py:1477]: It's interesting idea. 
I'd use another either use another name (`search_count_silent` ?) or keep the name, but return dictionary instead (e.g. `{recordCount: 0, isValid: true/false }`).

**Review by yelizariev — APPROVED** [review]

**Review by Julien00859 — CHANGES_REQUESTED** [review]
The only correct way to solve this problem is to *some-how* reraise the ValueError as a UserError as this domains comes from the UI. Adding a new method (in stable!!!) on models.py for the sole purpose of hiding exceptions (!!!) is a bad way to do it. Don't change the ValueError to a UserError in `expression.py`, most of the time it is a programming error thus ValueError is the right way to raise the exception.

In the current (master) information, the user receives an error message when the domain is invalid. The only bad thing is that that error is reported as a "server error" although it is the user who screwed up.

The implementation in your branch entirely hide the problem to the client, he doesn't receive any message whatsoever explaining that his domain is bad and that he should correct it. User don't have access to logs.

**Review by rco-odoo — CHANGES_REQUESTED** [review]
I totally second @Julien00859 's remark.
One has to distinguish between _user errors_ and _programming errors_.
User errors are not server errors, they are the expected way for the server to handle the request.  Programming errors are the ones you should see on sentry.

**Comment by Julien00859** [comment]
The domain should be validated upon saving the form

**Review by Julien00859 — CHANGES_REQUESTED** [review]
You didn't understand my point. Here you always reraise ValueError as UserError no matter if it is a programming or a user error. Meaning that if I'm a developer and I code a search with a bad domain, I receive a UserError instead of a ValueError.

Can you please stop changing the framework? The framework is not to blame here. The problem is not in the traceback either, it is a symptom of a bad domain saved on a record. Your commit message says that the problem is related to saving a bad domain in mass mailing, why am I still reading changes to `models.py` and not to mass mailing??

**Comment by sauc-odoo** [comment]
Hello @Julien00859 
The traceback will be generated from each and every place where `widget="domain"` is used (check here - 
 (odoo)https://github.com/search?q=repo%3Aodoo%2Fodoo%20widget%3D%22domain%22&type=code (enterprise)https://github.com/search?q=repo%3Aodoo%2Fenterprise%20widget%3D%22domain%22&type=code) and we will enter the invalid domain(for example: `[('', '=', 0)]`).

**Comment by Julien00859** [comment]
Then we must fix each and every field for which that widget is used or find a generic solution for them all. You shouldn't be allowed to save a bad domain.

**Comment by sauc-odoo** [comment]
I have created a draft PR- https://github.com/odoo/odoo/pull/125139 in master as POC to solve this issue.
I have implied two approaches:
1. I have reduced an RPC call to get recordCount.
2. If the user gives a bad domain a ValidationError will be raised.

I have achieved these results by creating a compute field `record_search_count` to get record count and that compute method will raise a ValidationError if a bad domain is added.

Kindly give your feedback on this approach.

**Comment by sauc-odoo** [comment]
The issue has been fixed in following PR:
https://github.com/odoo/odoo/pull/112326
https://github.com/odoo/odoo/pull/127699

**Pull request closed without merging**